File: javascript/packages/linter/src/rules/actionview-prefer-collection-render.ts
import { ParserRule } from "../types.js"
import { BaseRuleVisitor } from "./rule-utils.js"
import { isERBOutputNode, isNode, HTMLTextNode } from "@herb-tools/core"

import type { ERBIterationBlockNode, ERBRenderNode, Node, ParseResult, ParserOptions } from "@herb-tools/core"
import type { FullRuleConfig, LintContext, UnboundLintOffense } from "../types.js"

class PreferCollectionRenderVisitor extends BaseRuleVisitor {
  visitERBIterationBlockNode(node: ERBIterationBlockNode): void {
    this.checkIterationBlock(node)

    this.visitChildNodes(node)
  }

  private checkIterationBlock(node: ERBIterationBlockNode): void {
    if (node.message?.value !== "each") return

    const receiver = node.receiver?.value
    if (!receiver) return

    const blockArgument = this.singleBlockArgumentName(node)
    if (!blockArgument) return

    const render = this.singleRenderNodeInBody(node)
    if (!render) return

    const suggestion = this.collectionRenderFor(render, receiver, blockArgument)
    if (!suggestion) return

    this.addOffense(
      `Prefer \`${suggestion}\` over rendering a partial once per iteration. Collection rendering builds the partial once instead of for every item.`,
      node.location,
    )
  }

  private singleBlockArgumentName(node: ERBIterationBlockNode): string | null {
    if (node.block_arguments.length !== 1) return null

    const argument = node.block_arguments[0] as { kind?: string; name?: { value?: string } }

    if (argument.kind !== "positional") return null

    return argument.name?.value ?? null
  }

  private singleRenderNodeInBody(node: ERBIterationBlockNode): ERBRenderNode | null {
    const body = node.body.filter(child => !this.isWhitespaceOnlyText(child))

    if (body.length !== 1) return null

    const [child] = body

    if (child.type !== "AST_ERB_RENDER_NODE") return null

    const render = child as ERBRenderNode

    return isERBOutputNode(render) ? render : null
  }

  private isWhitespaceOnlyText(node: Node): boolean {
    return isNode(node, HTMLTextNode) && node.content.trim() === ""
  }

  private collectionRenderFor(render: ERBRenderNode, receiver: string, blockArgument: string): string | null {
    const keywords = render.keywords

    if (!keywords) return null
    if (keywords.object?.value === blockArgument) return `<%= render ${receiver} %>`

    const partial = keywords.partial?.value
    if (!partial) return null

    const locals = keywords.locals ?? []
    if (locals.length !== 1) return null

    const [local] = locals as Array<{ name?: { value?: string }, value?: { content?: string } }>
    if (local.value?.content !== blockArgument) return null

    const localName = local.name?.value
    if (!localName) return null

    const as = localName === this.collectionLocalNameFor(partial) ? "" : `, as: :${localName}`

    return `<%= render partial: "${partial}", collection: ${receiver}${as} %>`
  }

  private collectionLocalNameFor(partial: string): string {
    const name = partial.split("/").pop() ?? partial

    return name.startsWith("_") ? name.slice(1) : name
  }
}

export class ActionViewPreferCollectionRenderRule extends ParserRule {
  static ruleName = "actionview-prefer-collection-render"
  static introducedIn = this.version("unreleased")

  get defaultConfig(): FullRuleConfig {
    return {
      enabled: true,
      severity: "error"
    }
  }

  get parserOptions(): Partial<ParserOptions> {
    return {
      iteration_nodes: true,
      render_nodes: true,
    }
  }

  check(result: ParseResult, context?: Partial<LintContext>): UnboundLintOffense[] {
    const visitor = new PreferCollectionRenderVisitor(this.ruleName, context)

    visitor.visit(result.value)

    return visitor.offenses
  }
}

File: rust/herb-linter/src/rules/actionview_prefer_collection_render.rs
use herb::nodes::{AnyNode, ERBIterationBlockNode, ERBRenderNode};
use herb::Visitor;

rule_visitor!(PreferCollectionRenderVisitor);
define_parser_rule!(
  ActionViewPreferCollectionRenderRule,
  "actionview-prefer-collection-render",
  Error,
  PreferCollectionRenderVisitor,
  parser_options: { iteration_nodes: true, render_nodes: true },
  introduced_in: "unreleased"
);

fn single_block_argument_name(node: &ERBIterationBlockNode) -> Option<&str> {
  if node.block_arguments.len() != 1 {
    return None;
  }

  match &node.block_arguments[0] {
    AnyNode::RubyParameterNode(parameter) if parameter.kind == "positional" => parameter.name.as_ref().map(|token| token.value.as_str()),
    _ => None,
  }
}

fn single_render_node_in_body(node: &ERBIterationBlockNode) -> Option<&ERBRenderNode> {
  let mut body = node.body.iter().filter(|child| !is_whitespace_only_text(child));

  let child = body.next()?;

  if body.next().is_some() {
    return None;
  }

  match child {
    AnyNode::ERBRenderNode(render) if is_erb_output_render(render) => Some(render),
    _ => None,
  }
}

fn is_whitespace_only_text(node: &AnyNode) -> bool {
  match node {
    AnyNode::HTMLTextNode(text) => text.content.trim().is_empty(),
    _ => false,
  }
}

fn is_erb_output_render(node: &ERBRenderNode) -> bool {
  node
    .tag_opening
    .as_ref()
    .map(|token| token.value == "<%=" || token.value == "<%==")
    .unwrap_or(false)
}

fn collection_local_name_for(partial: &str) -> &str {
  let name = partial.rsplit('/').next().unwrap_or(partial);

  name.strip_prefix('_').unwrap_or(name)
}

fn collection_render_for(render: &ERBRenderNode, receiver: &str, block_argument: &str) -> Option<String> {
  let keywords = render.keywords.as_ref()?;

  if keywords.object.as_ref().map(|token| token.value.as_str()) == Some(block_argument) {
    return Some(format!("<%= render {receiver} %>"));
  }

  let partial = keywords.partial.as_ref().map(|token| token.value.as_str())?;

  if keywords.locals.len() != 1 {
    return None;
  }

  let local = match &keywords.locals[0] {
    AnyNode::RubyRenderLocalNode(local) => local,
    _ => return None,
  };

  if local.value.as_ref().map(|value| value.content.as_str()) != Some(block_argument) {
    return None;
  }

  let local_name = local.name.as_ref().map(|token| token.value.as_str())?;
  let as_name = if local_name == collection_local_name_for(partial) {
    String::new()
  } else {
    format!(", as: :{local_name}")
  };

  Some(format!("<%= render partial: \"{partial}\", collection: {receiver}{as_name} %>"))
}

impl Visitor for PreferCollectionRenderVisitor {
  fn visit_erb_iteration_block_node(&mut self, node: &ERBIterationBlockNode) {
    if let Some(suggestion) = suggestion_for(node) {
      self.add_offense(
        format!("Prefer `{suggestion}` over rendering a partial once per iteration. Collection rendering builds the partial once instead of for every item."),
        node.location.clone(),
      );
    }

    self.walk_erb_iteration_block_node(node);
  }
}

fn suggestion_for(node: &ERBIterationBlockNode) -> Option<String> {
  if node.message.as_ref().map(|token| token.value.as_str()) != Some("each") {
    return None;
  }

  let receiver = node.receiver.as_ref().map(|token| token.value.as_str())?;
  let block_argument = single_block_argument_name(node)?;
  let render = single_render_node_in_body(node)?;

  collection_render_for(render, receiver, block_argument)
}
